training/keras_training.py

```diff
@@ -213,30 +264,34 @@ def main(args, config):
         recommend_steps_per_epoch=int(min([len(classIndexDict[class_]) for class_ in classes])/(batch_size/len(classes)))+1
         logger.info("steps_per_epoch: Using {} instead of recommended minimum of {}".format(str(steps_per_epoch), str(recommend_steps_per_epoch)))
     else:
-        logger.info("model: steps_per_epoch: Not found in {} ",args.config)
+        logger.info("model: steps_per_epoch: Not found in {} ".format(args.config))
         raise Exception
 
     model_impl = getattr(keras_models, config["model"]["name"])
-    model = model_impl(len(variables), len(classes))
+    model = model_impl(len(variables) + len_eras, len(classes))
     model.summary()
     if(args.balance_batches):
         logger.info("Running on balanced batches.")
+        eraIndexDict={era: {label: np.where((z_train[:,i_era] == 1) & (y_train[:,i_class] == 1))[0] for i_class, label in enumerate(classes)} for i_era, era in enumerate(eras)}
         def balancedBatchGenerator(batch_size):
             while True:
-                nperClass=int(batch_size/len(classes))
-                selIdxDict={label:classIndexDict[label][np.random.randint(0,len(classIndexDict[label]),nperClass)] for label in classes}
-                y_collect=np.concatenate([y_train[selIdxDict[label]] for label in classes])
-                x_collect=np.concatenate([x_train[selIdxDict[label],:] for label in classes])
-                w_collect=np.concatenate([w_train[selIdxDict[label]]*(batch_size/np.sum(w_train[selIdxDict[label]])) for label in classes])
+                nperClass=int(batch_size/(len(classes)*len(eras)))
+                selIdxDict={era: {label:eraIndexDict[era][label][np.random.randint(0,len(eraIndexDict[era][label]),nperClass)] for label in classes} for era in eras}
+                y_collect=np.concatenate([y_train[selIdxDict[era][label]] for label in classes for era in eras])
+                x_collect=np.concatenate([x_train[selIdxDict[era][label],:] for label in classes for era in eras])
+                w_collect=np.concatenate([w_train[selIdxDict[era][label]]*(batch_size/np.sum(w_train[selIdxDict[era][label]])) for label in classes for era in eras])
                 yield x_collect, y_collect,w_collect
 
         def calculateValidationWeights(x_test, y_test, w_test):
-            testIndexDict = {label: np.where(y_test[:, i_class] == 1)[0] for i_class, label in enumerate(classes)}
-            y_collect = np.concatenate([y_train[testIndexDict[label]] for label in classes])
-            x_collect = np.concatenate([x_train[testIndexDict[label], :] for label in classes])
-            ## len(x_test) for moving the validation loss to a useful scale
+            testIndexDict = {
```

Comment:
defines the indices of the correct answers  for the each era / class combination